### ISSUE TITLE
Moving a subdirectory in shared volume gives the container root access to the whole host filesystem

### ISSUE THREAD

**Issue by lzlarryli** [docker/docker#14890]
Moving a subdirectory in a shared volume on the host when the container is in that subdirectory gives the container root access to the entire host filesystem. To replicate this, run on the host:

    mkdir -p ~/tmp/foo/bar
    docker run -it --rm --volume ~/tmp/foo:/opt ubuntu:latest

In the container,
 
    cd /opt/bar

Then in the host,

    mv ~/tmp/foo/bar ~/tmp/

Then in the container,

    cd ..

After this, the prompt in the container becomes:

    root@eb9671bb0bd9:/opt/home/larry/tmp# 

Then the container gets full root access to the entire host filesystem via relative path, that is, for example in the above `cd /opt/usr/bin` does not work, but `cd ../../../usr/bin` works. A malicious user in the container can for example rm everything on the host. We also note that in step 3, if `~/tmp/foo` is moved, then everything is fine and we do not have this problem.

It seems to me that at least in the documentation it should mention that shared directories should not be touched on the host when a live container is using it.

Relevant information:

```
[22:53 ~]$ uname -a
Linux quetzalcoatlus 4.1.2-2-ARCH #1 SMP PREEMPT Wed Jul 15 08:30:32 UTC 2015 x86_64 GNU/Linux
[22:53 ~]$ docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64
[22:53 ~]$ docker info
Containers: 3
Images: 47
Storage Driver: devicemapper
 Pool Name: docker-254:0-6297448-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 6.561 GB
 Data Space Total: 107.4 GB
 Data Space Available: 88.43 GB
 Metadata Space Used: 6.083 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.141 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.100 (2015-06-30)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.2-2-ARCH
Operating System: Arch Linux
CPUs: 8
Total Memory: 23.44 GiB
Name: quetzalcoatlus
ID: 6OP3:QZ3Z:65WE:45AV:7V3N:LV4N:ISFL:JN43:HOL3:4OM5:7OND:3CBX
```

**Comment by ewindisch** [comment]
Thank you for your report. We do request that such reports are sent to security@docker.com in spirit of responsible disclosure.

In this case, we are aware of this issue, but it requires a malicious user on the outside of the container with either root access, or with permission to modify files and directories utilized by a root user. All users in the Docker group are, by this definition, "root users" given the elevated privileges provided all users with access to the API.

This is a solid candidate for documentation as a best-practice. It may also be something preventable using SELinux.

**Comment by thaJeztah** [comment]
@ewindisch this is a dup of https://github.com/docker/docker/issues/12317, so I think this should be closed. If you think this should be documented, could you create a separate issue (or a PR) for that?

**Comment by thaJeztah** [comment]
@lzlarryli do you agree with closing (and the explanation @ewindisch gave here and in https://github.com/docker/docker/issues/12317)? Or do you have additional thoughts?

**Comment by lzlarryli** [comment]
@thaJeztah I agree with the explanation @ewindisch. Since it is stated that this issue is known, is not considered a bug, and will not be fixed, I do not see the value of having this issue open. I think the current design that "docker group ~ root" severely restricts docker's use, including for my use case. But consider that design as given, arguments @ewindisch are perfectly fine, especially given that he agreed that it should be documented. 

I do no know where to submit a request for change of documentation for docker. To my understanding, it is not on Github. It would be nice if @thaJeztah or @ewindisch can give a pointer.

**Comment by thaJeztah** [comment]
@lzlarryli good to hear, I wanted to check with you because you closed without commenting (we had some "angry" users in the past, and that's not our goal :smile:)

Actually, you *can* create requests for the documentation in this repository. The documentation can be found here: https://github.com/docker/docker/tree/master/docs and the man pages are https://github.com/docker/docker/tree/master/man. All documentation is generated from those Markdown files.

If you you have ideas on how this could be improved in the documentation, you could even consider creating a pull request (we really welcome those). The pull request doesn't have to be "perfect" directly, there will still be a review and the changes can be discussed there.

**Comment by lzlarryli** [comment]
@thaJeztah Hehe, there is no need to worry~ I was in a hurry and closed the issue before I commented on it. Then I realized that the "issuing closing line" does not have an "edit" button on it and thought it is probably an unnecessary escalation to reopen the issue to add the comment:-)

I am checking out how to prevent this using SELinux and will contribute to the documentation somewhere this week~

**Comment by thaJeztah** [comment]
> I am checking out how to prevent this using SELinux and will contribute to the documentation somewhere this week~

oh, nice! thanks in advance!

**Comment by lzlarryli** [comment]
I added a PR for the documentation for this. For my use case, I found that I can either add sticky bits to the shared directory or `chattr +a` to make the directory tree structure immutable as a dirty workaround.  

In the long run, a bit of maintainence on the host like renaming directories might expose the entire host filesystem to the container still sound pretty bad. Is this part implemented by docker or part of the linux kernel?
